contracts/floaty/src/instructions.rs

```diff
@@ -0,0 +1,551 @@
+    };
+    ("f32.convert_i32_u", $input:expr) => {
+        $crate::instructions::run_instr!("f32.convert_i32_u", $input, u32, f32)
+    };
+    ("f32.convert_i64_s", $input:expr) => {
+        $crate::instructions::run_instr!("f32.convert_i64_s", $input, i64, f32)
+    };
+    ("f32.convert_i64_u", $input:expr) => {
+        $crate::instructions::run_instr!("f32.convert_i64_u", $input, u64, f32)
+    };
+    ("f32.demote_f64", $input:expr) => {
+        $crate::instructions::run_instr!("f32.demote_f64", $input, f64, f32)
+    };
+    ("f64.convert_i32_s", $input:expr) => {
+        $crate::instructions::run_instr!("f64.convert_i32_s", $input, i32, f64)
+    };
+    ("f64.convert_i32_u", $input:expr) => {
+        $crate::instructions::run_instr!("f64.convert_i32_u", $input, u32, f64)
+    };
+    ("f64.convert_i64_s", $input:expr) => {
+        $crate::instructions::run_instr!("f64.convert_i64_s", $input, i64, f64)
+    };
+    ("f64.convert_i64_u", $input:expr) => {
+        $crate::instructions::run_instr!("f64.convert_i64_u", $input, u64, f64)
+    };
+    ("f64.promote_f32", $input:expr) => {
+        $crate::instructions::run_instr!("f64.promote_f32", $input, f32, f64)
+    };
+    //
+    ("i32.reinterpret_f32", $input:expr) => {
+        $crate::instructions::run_instr!("i32.reinterpret_f32", $input, f32, i32)
+    };
+    ("i64.reinterpret_f64", $input:expr) => {
+        $crate::instructions::run_instr!("i64.reinterpret_f64", $input, f64, i64)
+    };
+    ("f32.reinterpret_i32", $input:expr) => {
+        $crate::instructions::run_instr!("f32.reinterpret_i32", $input, u32, f32)
+    };
+    ("f64.reinterpret_i64", $input:expr) => {
+        $crate::instructions::run_instr!("f64.reinterpret_i64", $input, u64, f64)
+    };
+    //
+    ("i32.trunc_sat_f32_s", $input:expr) => {
+        $crate::instructions::run_instr!("i32.trunc_sat_f32_s", $input, f32, i32)
+    };
+    ("i32.trunc_sat_f32_u", $input:expr) => {
+        $crate::instructions::run_instr!("i32.trunc_sat_f32_u", $input, f32, u32)
+    };
+    ("i32.trunc_sat_f64_s", $input:expr) => {
+        $crate::instructions::run_instr!("i32.trunc_sat_f64_s", $input, f64, i32)
+    };
+    ("i32.trunc_sat_f64_u", $input:expr) => {
+        $crate::instructions::run_instr!("i32.trunc_sat_f64_u", $input, f64, u32)
+    };
+    ("i64.trunc_sat_f32_s", $input:expr) => {
+        $crate::instructions::run_instr!("i64.trunc_sat_f32_s", $input, f32, i64)
+    };
+    ("i64.trunc_sat_f32_u", $input:expr) => {
+        $crate::instructions::run_instr!("i64.trunc_sat_f32_u", $input, f32, u64)
+    };
+    ("i64.trunc_sat_f64_s", $input:expr) => {
+        $crate::instructions::run_instr!("i64.trunc_sat_f64_s", $input, f64, i64)
+    };
+    ("i64.trunc_sat_f64_u", $input:expr) => {
+        $crate::instructions::run_instr!("i64.trunc_sat_f64_u", $input, f64, u64)
+    };
+}
+pub(crate) use run;
+
+#[cw_serde]
+pub enum Value {
+    U32(u32),
+    U64(u64),
+    F32(u32),
+    F64(u64),
+}
```

Comment:
Do I understand correctly that in JSON this looks something like this?

```json
{"f64":16048667344127309905}
```

So it's well supported for all hosts that handles 64 bit integers correctly but we don't risk any loss in precision?

Reply:
Yes, that was the intention behind this. It also makes sure that different NaNs would produce different outputs.

Reply:
Great, then this should work well in a Go environment too which supports int64/uint64 for JSON.